cuML/src/glm/qn/simple_mat.h

```diff
@@ -294,55 +294,48 @@ std::ostream &operator<<(std::ostream &os, const SimpleMat<T> &mat) {
   return os;
 }
 
-// TODO The following classes own their storage and are just for testing
-// The internals should be replaced by device_buffer+device_allocator, when ready
 template <typename T> struct SimpleVecOwning : SimpleVec<T> {
   typedef SimpleVec<T> Super;
-  bool allocated;
+  typedef MLCommon::device_buffer<T> Buffer;
+  std::unique_ptr<Buffer> buf;
 
-  SimpleVecOwning() : Super(), allocated(false) {}
+  SimpleVecOwning() : Super() {}
 
-  SimpleVecOwning(int n) : Super(), allocated(false) { reset(n); }
-
-  ~SimpleVecOwning() noexcept(false) { CUDA_CHECK(cudaFree(Super::data)); }
+  SimpleVecOwning(const cumlHandle_impl &handle, int n, cudaStream_t stream)
+      : Super() {
+    reset(handle, n, stream);
+  }
 
-  void reset(int n) {
-    if (allocated) {
-      CUDA_CHECK(cudaFree(this->data));
-    }
-    MLCommon::allocate(this->data, n);
-    Super::reset(this->data, n);
-    allocated = true;
+  void reset(const cumlHandle_impl &handle, int n, cudaStream_t stream) {
+    buf.reset(new Buffer(handle.getDeviceAllocator(), stream, n));
+    Super::reset(buf->data(), n);
   }
 
-  void operator=(const SimpleVec<T> &other) { Super::operator=(other); }
+  void operator=(const SimpleVec<T> &other) = delete;
 };
 
 template <typename T> struct SimpleMatOwning : SimpleMat<T> {
   typedef SimpleMat<T> Super;
-  bool allocated;
+  typedef MLCommon::device_buffer<T> Buffer;
+  std::unique_ptr<Buffer> buf;
   using Super::m;
   using Super::n;
   using Super::ord;
 
-  SimpleMatOwning(STORAGE_ORDER order = COL_MAJOR)
-      : Super(order), allocated(false) {}
+  SimpleMatOwning(STORAGE_ORDER order = COL_MAJOR) : Super(order) {}
 
-  SimpleMatOwning(int m, int n, STORAGE_ORDER order = COL_MAJOR)
-      : Super(order), allocated(false) {
-    reset(m, n);
+  SimpleMatOwning(const cumlHandle_impl &handle, int m, int n,
```

Comment:
Why not only pass in a `std::shared_ptr<device_allocator>` instead of `cumlHandle_impl`?